poetry/puzzle/provider.py

```diff
@@ -710,7 +710,7 @@ def debug(self, message: str, depth: int = 0) -> None:
 
         if message.startswith("fact:"):
             if "depends on" in message:
-                m = re.match(r"fact: (.+?) depends on (.+?) \((.+?)\)", message)
+                m = re.match(r"fact: (.+?) depends on ((.+?)($| \((.+?)\)))", message)
```

Comment:
Since we reference `m.group(2)` and `m.group(3)`, this require more changes.

```suggestion
                m = re.match(r"fact: (.+?) depends on (?:(.+?)($| \((.+?)\)))", message)
```

And we need to change the way we format the message in case `m.group(3)` is empty.

Reply:
With the regex as I made it m.group(2) == m.group(3).

So `message == "fact: xxx (==1.7.0) depends on cryptography"`, group 2 and 3 would be `cryptography`.

I have changed to match your request, if group 3 is `None` I put a `*`.  Not sure what you'd like to see here.

I think having group 2 and 3 be equal in this case might be the most logical solution for what the debug log prints.